### ISSUE TITLE
Bad initialization/no zero cluster handling will lead to low utilization and eventually low accuracy [KMeans seems to be broken]

### ISSUE THREAD

**Issue by CanYing0913** [jokofa/torch_kmeans#12]
It turns out, when online K-means (i.e. when we need to fit the K-means model multiple times to let it go through a very large dataset) is used, it is more likely the clustered centers will contain many zero centers. It will lead to low utilization and low accuracy at the end, since it wastes many clusters on 0s and thus, underrepresents the entire dataset.

The cause of the issue is: Since the second iteration of the online K-means, after the initial centers are initialized, the assigned indices may lead to some clusters not containing any data points. And because of this library did not handle this scenario, it will create multiple "holes" for those empty clusters. As more iterations we go, more holes will add up. Eventually, it is highly possible to have a codebook of 32 clusters, for example, with only one to two valid centers.

I took a look at `scikit-learn` to see how they handle this case. And it seems they are randomly initializing the zero centers again.

-----Previous Problem Description-------
I highly suspect the modules are broken right now. I used this package for my own project, but they always end up with bad result. So I ran a couple of tests the other day, and made an interesting observation:
1. Most of the center entries are empty: I am using 32 as cluster number, 1 as `num_init`, and 10 as iterations. I use provided `forward` function (thanks for doing that btw) to smoothly train it over a large dataset in a loop since I cannot load those data in one pass due to memory capacity limitations. When I tried to inspect what are inside of the cluster tables, I found two things: 
    - First, the labels (i.e. the indices pointing to the closest clusters) are well-patterned, for example:
    `labels = tensor([1, 1, 1, 1, 1, 1, 1, 1, 1, 5, 5, 5, 1, 1, 1, 1, 1, 1, 1, 1, 1])` 
    I am curious as why the labels unanimously vote for cluster 1 and 5, so I took a look at the clusters, and made my second discovery:
    - Second, the clusters other than those within the labels, are all 0s.
    And I already checked my data to see if that is a 0 pattern within my dataset. The dataset is indeed a valid dataset with mostly non-zero input.
2. The previous one happens to me under the scenario of having many (>100) Kmeans instance in parallel. I do not know if that is the cause of the issue, but it certainly should not behave like this.

Edit: It seems a single fit is still working. If we fit it multiple times, it will eventually end up with the problem described above.

**Comment by VoVAllen** [comment]
How did you patch the code to solve the issue?

**Comment by CanYing0913** [comment]
@VoVAllen I didn't get the time to do so, but it is a little **tricky** to deal with when we randomly generate new clusters. The most probable result will not make a difference: those random new clusters are likely still not being assigned with any data points, so it will need more thinking here. My thought is we may need to randomly find a point in dataset range and re-assign clusters before the next iteration, but then it is pretty similar to initialization problem - `kmeans++` is already an implementation of the idea mentioned above.

**Comment by CanYing0913** [comment]
#13 . Also it should be dependent on the number of centers we choose, so this issue is invalid anymore.
